### ISSUE TITLE
No way for low-overhead total norm in native PyTorch with large number of tensors

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#133586]
**Context**
Gradient norm clipping is a popular technique for stabilizing training, which requires computing the total norm with respect to the model's gradients. This involves a norm reduction across all of the gradients down to a single scalar.

PyTorch's `clip_grad_morm_` offers both single-tensor (`foreach=False`) and multi-tensor (`foreach=True`) implementations. However, even the multi-tensor `foreach` implementation incurs high CPU overhead when computing the total norm over a large number of tensors.
https://github.com/pytorch/pytorch/blob/3434a54fba537f67d152b50142b211d4aa059e67/torch/nn/utils/clip_grad.py#L81-L96
The foreach implementation involves:
1. `torch._foreach_norm(tensors)` to return a list of 0D scalars, representing the norm of each gradient
2. `torch.stack` to cat the 0D scalars into a 1D tensor
3. `torch.linalg.vector_norm()` to compute the norm of the norms, representing the final total norm

**Issue**
The foreach implementation incurs much unnecessary CPU overhead, making the clipping heavily CPU bound when operating on a large number of tensors. For example, for ~2000 tensors, the total norm calculation takes >18 ms on CPU, while only 1.3 ms on GPU (for a particular real workload -- larger tensors would make this slower).
![Screenshot 2024-08-15 at 12 47 18 PM](https://github.com/user-attachments/assets/544213f7-a558-4391-a023-f179ac0e8118)

Assuming `N` tensors, some inefficiencies in the existing implementation arise from:
- `_foreach_norm` must call `N` `aten::empty({})` to construct the `N` 0D scalar outputs ([code](https://github.com/pytorch/pytorch/blob/bedf96d7ffe74b34bcfe52c7ae1ae05f40d6c8ee/aten/src/ATen/native/cuda/ForeachReduceOp.cu#L444)).
- The `N` 0D scalars need to be `stack`d for the final norm reduction. `stack` requires 1D tensors, so each of the `N` 0D scalar gets unsqueezed again in `stack`.

Together, this leads to an extra `2N` dispatcher calls to handle the `N` intermediate scalars. Ideally, we can avoid materializing these `N` intermediates, especially as `torch.Tensor`s, where one option is a fused kernel.

Today, `torch.compile` cannot address this issue in a satisfying way. Default `torch.compile` cannot achieve horizontal fusion, leading to slower performance than eager mode. `torch.compile(mode="reduce-overhead")` does reduce overhead more than eager but results in 2x memory usage, likely due to copying the gradients into CUDA graph addresses. Note that we likely cannot mark the inputs as static because gradients are computed anew every iteration at possibly different addresses, and for my use case, we cannot compile the gradient allocation with `torch.compile` (due to FSDP).

Here is an example script for getting profiler traces of various implementations: P1529496302

The ask is to provide _some_ native way to make this total norm calculation not CPU-overhead bound, e.g. via a fused op.

**Comment by janeyx99** [comment]
cc @mlazos This looks similar to compiling across foreach ops for the horizontal fusion portion. Is this something that torch.compile could fuse?

**Comment by carmocca** [comment]
Since #139467 separated the norm computation from the clipping, the following is no longer a limitation for all cases

> Note that we likely cannot mark the inputs as static because gradients are computed anew every iteration at possibly different addresses

As [`_get_total_norm`](https://github.com/pytorch/pytorch/blob/ce80865f134c621051afc7e30e4a71cacf3cfaf1/torch/nn/utils/clip_grad.py#L44-L112) can be called for non-grad tensors too (like parameters or optimizer states)

So the huge cpu overhead of the norm calculation is highly relevant in this case, because computing the norms for those objects needs to happen after the optimizer step so there are few opportunities to overlap the work within the same training step. It would need to be overlapped with the next one.

**Comment by mlazos** [comment]
You should be able to use `foreach_map` to compile a fast kernel to do this. Lmk if you need help with this. There is an example implementing adam with it in the [compiled optimizer tests](https://github.com/pytorch/pytorch/blob/4ab967c44d4b267f6d20a3235e30052cc823ee07/test/inductor/test_compiled_optimizers.py#L933).

**Comment by awgu** [comment]
@mlazos can you clarify which part of the problem `foreach_map` is addressing (e.g. compared to existing `_foreach_norm`)?

Also, looking at the unit test, should we be using the output of this `foreach_map`? Or can you explain the in-place semantic if not?
https://github.com/pytorch/pytorch/blob/4ab967c44d4b267f6d20a3235e30052cc823ee07/test/inductor/test_compiled_optimizers.py#L114-L115

**Comment by awgu** [comment]
@mlazos I tried some variations of using `foreach_map` with `torch.linalg.vector_norm`, and they error:
```
LoweringException: SubgraphLoweringException: Buffers cannot be created while lowering a pointwise subgraph. This could be for a good reason (e.g. you're calling an op we can't codegen as a pointwise op), but it could also be a bug. Please file a bug report if you think this should be supportable.
```

Here is a minimal usage:
<details>
<summary> Repro </summary>

```
import torch

from torch._higher_order_ops import foreach_map

def compute_total_norm(tensors: list[torch.Tensor], norm_type: float):
    norms = foreach_map(torch.linalg.vector_norm, (tensors,), ord=norm_type)
    return torch.linalg.vector_norm(torch.stack(norms), ord=norm_type)


if __name__ == "__main__":
    torch.manual_seed(42)
    num_tensors = 10
    ts = [
        torch.randn((1024, 1024), device="cuda") for _ in range(num_tensors)
    ]

    @torch.compile()
    def compute_total_norm_fn():
        return compute_total_norm(ts, norm_type=2.0)

    total_norm = compute_total_norm_fn()
```
</details>

**Comment by carmocca** [comment]
I noticed that `_foreach_norm()` can be `@torch.compiled`. However, it's not faster as per my benchmarks:

```python
import torch
from torch.nn.utils.clip_grad import _get_total_norm
from triton.testing import do_bench

n = 2.0
ts = [torch.randn((1024, 1024), device="cuda") for _ in range(2000)]

def torch_noforeach():
    return _get_total_norm(ts, norm_type=n, foreach=False)
    
def torch_foreach():
    return _get_total_norm(ts, norm_type=n, foreach=True)

@torch.compile(fullgraph=True, backend="cudagraphs")
def compiled_noforeach():
    ns = torch.stack([torch.linalg.vector_norm(t, n) for t in ts])
    return torch.linalg.vector_norm(ns, ord=n)

@torch.compile(fullgraph=True, backend="cudagraphs")
def compiled_foreach():
    ns = torch.stack(torch._foreach_norm(ts, n))
    return torch.linalg.vector_norm(ns, ord=n)

import amp_C
from apex.multi_tensor_apply import MultiTensorApply
multi_tensor_applier = MultiTensorApply(2048*32*4)

def apex_l2():
    dummy_overflow_buf = torch.zeros([1], dtype=torch.int32, device="cuda")
    ns = [
        multi_tensor_applier(
            amp_C.multi_tensor_l2norm,
            dummy_overflow_buf,
            [ts],
            False,
        )[0]
    ]
    return torch.linalg.vector_norm(torch.stack(ns), ord=n)

"""
expected = torch_noforeach()
torch.testing.assert_close(torch_foreach(), expected)
torch.testing.assert_close(compiled_noforeach(), expected)
torch.testing.assert_close(compiled_foreach(), expected)
torch.testing.assert_close(apex_l2(), expected)
"""

for f in (
    torch_noforeach,
    torch_foreach,
    compiled_noforeach,
    compiled_foreach,
    apex_l2,
):
    ms = do_bench(f, warmup=100, rep=1000)
    print(f.__name__, ms)
```

```text
torch_noforeach 23.528676986694336
torch_foreach 4.801433086395264
compiled_noforeach 22.68472671508789
compiled_foreach 7.207825660705566
apex_l2 2.155019760131836
```

**Comment by mlazos** [comment]
@carmocca fwiw, setting backend to `"inductor"` and using `mode="reduce-overhead"` gives way better numbers at least wrt to CPU overhead ([script](https://gist.github.com/mlazos/7cab5840840fac25f31e59680865ea8c))

GPU:
torch_noforeach 20.461780548095703
torch_foreach 4.9132466316223145
compiled_noforeach 20.559999465942383
compiled_foreach 16.73971176147461
CPU:
torch_noforeach 26.810669000000004
torch_foreach 6.073393000000001
compiled_noforeach 25.078167
compiled_foreach 21.125101

re: foreach_map I misunderstood and thought that you weren't getting any horizontal fusion, which is why I suggested foreach_map. Since there is already a foreach_norm op, this should force the compiler to horizontally fuse.
One potential improvement for compile here is to allow eliding the N 0d tensors,  and basically copying directly into one buffer for the final reduce. I can profile this on tuesday and see if we are doing that today.

**Comment by awgu** [comment]
When I compile:
```
def compute_total_norm(tensors: list[torch.Tensor], norm_type: float):
    norms = torch._foreach_norm(tensors, ord=norm_type)
    return torch.linalg.vector_norm(torch.stack(norms), ord=norm_type)
```
The default compiled version is slower than the eager version because the `stack` -> `vector_norm` is broken up into a fused copy and norm kernel _per tensor_, whereas in eager, there are a few cat kernels for `stack` and one norm kernel total.

**Comment by mlazos** [comment]
Hmm I just thought of one other way of doing this which might work. If you have torch._foreach_norm followed by _foreach_copy_ into a list of slices of a single buffer, it's possible inductor will fuse in this case. If not this will require something more in depth which I won't be able to do until I'm back from vacation March 3rd.
